Add support for location and scale tracks

armish opened the pull request:
This implements two tracks that do not use data sources, but create helper visualizations based on the visible genomic range (closes #62).

`LocationTrack` marks the middle of the visible range and shows the exact base location next to the line mark:
![screen recording 2015-07-23 at 10 32 pm](https://cloud.githubusercontent.com/assets/7809/8866614/836f1da0-318c-11e5-989e-ff6a23b0889a.gif)

`ScaleTrack` track includes a centered scale bar of which size is determined by the visible range and a `niceify` method that heuristically fine tunes the exact size of the bar. For regions bigger than 1000bp, the visualization creates a scale bar on the order of kilobase (kb), megabase (mb) or gigabase (gb):
![screen recording 2015-07-23 at 10 51 pm](https://cloud.githubusercontent.com/assets/7809/8866705/75ba7b0e-318d-11e5-8380-254884691d79.gif)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/pileup.js/241)
<!-- Reviewable:end -->

danvk commented:
Thanks for these tracks! I've got a few suggestions.

---

Review status: 0 of 8 files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/EmptySource.js, line 7 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-XnCLcIO_YNtZBOl9)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/EmptySource.js#L7)):</sup>
Is this necessary? I forget.

---

<sup>**[src/main/EmptySource.js, line 15 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-XlFSBcHlhfy4V20j)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/EmptySource.js#L15)):</sup>
two space indent

---

<sup>**[src/main/LocationTrack.js, line 35 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-YNJrixL10-xa7I0m)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L35)):</sup>
What do you think about drawing two vertical lines, one on the left edge of the center base and one on the right edge? It might be clearer what's going on when you're zoomed in. I believe IGV does something like this.

---

<sup>**[src/main/LocationTrack.js, line 41 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-YhCETq5fPhoI3zGs)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L41)):</sup>
Hello indentation!

---

<sup>**[src/main/LocationTrack.js, line 41 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-YjKCs6fQn9UTDTyn)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L41)):</sup>
Use a compact object literal: `labelSize: {height, width}`

---

<sup>**[src/main/LocationTrack.js, line 51 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-a56jKdts9L-DiTcO)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L51)):</sup>
I believe this is a built-in method on React classes—`getDOMNode()` should return the top-level element for this class.

---

<sup>**[src/main/LocationTrack.js, line 57 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-ZRgbGqoR4MSvRnhC)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L57)):</sup>
Not required, but if you wanted to you could also write this as:

```javascript
var {range, width, height} = this.props;
```

---

<sup>**[src/main/LocationTrack.js, line 63 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-Zl9SqFXVqzwCqfXz)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L63)):</sup>
Not strictly related to your change, but I'm curious if setting this on every frame (even when the height/width don't change) has an effect on performance.

---

<sup>**[src/main/LocationTrack.js, line 72 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-alz9AuexetBKT-X5)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L72)):</sup>
pull this up into a top-level constant

---

<sup>**[src/main/LocationTrack.js, line 74 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-azzj5qdHyaCN8oyR)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L74)):</sup>
I believe that if you set these styles on the text element, you can dispense with measuring it entirely:

```css
text-anchor: start;
dominant-baseline: middle;
```

---

<sup>**[src/main/LocationTrack.js, line 75 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-_CVINJh3BjV4kkmD)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L75)):</sup>
3?

---

<sup>**[src/main/LocationTrack.js, line 79 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-_WCAFKN5v9KOZSqM)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L79)):</sup>
you can write this more compactly as:

```javascript
midLine.attr({
  x1: midX,
  y1: 0,
  x2: midX,
  y2: height
});
```

(also below)

---

<sup>**[src/main/pileup.js, line 60 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-cJRq9D6caGDWmHV4)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/pileup.js#L60)):</sup>
I like that you don't have to explicitly set a `source` attribute for the Location/Scale tracks, but I do worry that this makes the API more error-prone. For example, what if I type `source` instead of `data` in the configuration for my pileup track? Previously it would have blown up but now it silently fails.

Maybe the visualizations should define a default data source that typically throws?

---

<sup>**[src/main/Root.js, line 12 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-d3eUKkWimjqy9N3N)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/Root.js#L12)):</sup>
This is unused in this module. (I'm surprised that's not a lint error, actually!)

---

<sup>**[src/main/ScaleTrack.js, line 2 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-dqGK-QOmu__rV2Ov)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/ScaleTrack.js#L2)):</sup>
an approximate ....?

---

<sup>**[src/main/ScaleTrack.js, line 55 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-gSozKyPhLxBV0NPb)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/ScaleTrack.js#L55)):</sup>
Maybe `formatRange`?

---

<sup>**[src/main/ScaleTrack.js, line 56 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-gMdpiocWSReB0hqG)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/ScaleTrack.js#L56)):</sup>
D3 has some built-in methods for doing this. See http://stackoverflow.com/a/14942605/388951

---

<sup>**[src/main/ScaleTrack.js, line 102 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-h7TKMYrF_V-5D_qW)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/ScaleTrack.js#L102)):</sup>
In IGV this line always spans the entire view. I think that's more important than having it be a "nice" number. It's good to know that you're looking at exactly a 3,931 base pair region.

---

<sup>**[src/main/ScaleTrack.js, line 105 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-hfe-fHo8U2CGCV9j)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/ScaleTrack.js#L105)):</sup>
It would be nice to have [arrowheads][1] on the scale line, like IGV does. (Feel free to file an issue for that.)

[1]: http://stackoverflow.com/a/12686815/388951

---

<sup>**[style/pileup.css, line 202 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-gg9N2ObQUBleQGPJ)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/style/pileup.css#L202)):</sup>
Maybe we should just set this globally for the pileup, we seem to be repeating it a lot:

```css
.pileup-root text {
   font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif;
}
```

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/241)
<!-- Sent from Reviewable.io -->

armish commented:
Review status: 0 of 8 files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/EmptySource.js, line 7 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-XnCLcIO_YNtZBOl9)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/EmptySource.js#L7)):</sup>
Yes and no. This piece of code was causing me trouble:
https://github.com/hammerlab/pileup.js/blob/master/src/main/Root.js#L49

I first thought about chaining a filter there with `.filter(t => t.source)` to prevent it calling the method from source-independent tracks. This worked but then the bundle started to complain about `VisualizationWrapper` not having a `source` property.

I thought this `EmptySource` might simplify things as a default option.

---

<sup>**[src/main/pileup.js, line 60 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-cJRq9D6caGDWmHV4)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/pileup.js#L60)):</sup>
that is a good point. I really don't want to force users to set the `data` field to `EmptySource` for this track, so that's why I went this way. 

As an alternative, tracks might suggest a default track via `getDefaultSource()` and we can use that to make sure the use of the empty source is on purpose. What do you think?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/241)
<!-- Sent from Reviewable.io -->

danvk commented:
Review status: 0 of 8 files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/EmptySource.js, line 7 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-XnCLcIO_YNtZBOl9)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/EmptySource.js#L7)):</sup>
Ah, sorry, I have no problems with `EmptySource` in general. I was just asking if the `type EmptySource` bit was strictly necessary. I think that it is.

---

<sup>**[src/main/pileup.js, line 60 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-cJRq9D6caGDWmHV4)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/pileup.js#L60)):</sup>
I think that's a great solution. We can either make `getDefaultSource` an optional method (in which case `PileupTrack` wouldn't provide it, since there's no sensible default) or make it required and put in simple implementations like this in `PileupTrack` and friends:

```javascript
  getDefaultSource() {
    throw "You must specify a data source for pileup visualizations"
  }
```

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/241)
<!-- Sent from Reviewable.io -->

armish commented:
Thanks a lot for all the constructive comments, Dan! I just addressed them all and updated the PR according to your comments.

---

Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/LocationTrack.js, line 35 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-YNJrixL10-xa7I0m)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L35)):</sup>
yeah, that was a good suggestion -- implemented this one.

---

<sup>**[src/main/LocationTrack.js, line 41 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-YhCETq5fPhoI3zGs)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L41)):</sup>
:( -- looks like it is time for me to start tweaking Atom.

---

<sup>**[src/main/LocationTrack.js, line 51 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-a56jKdts9L-DiTcO)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L51)):</sup>
Weird, but when you `extend` the `React.Component` class, that method cannot be accessed from the class, whereas if you do `React.createClass`, `getDOMNode()` works just great. 

Any ideas or suggestions for this?

---

<sup>**[src/main/LocationTrack.js, line 63 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-Zl9SqFXVqzwCqfXz)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L63)):</sup>
I tried to profile while resizing the window and nothing suspicious popped up to me.  We can re-factor the code for tracks to make sure we run this only when the size changes, should I create an issue for this?

---

<sup>**[src/main/LocationTrack.js, line 74 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-azzj5qdHyaCN8oyR)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L74)):</sup>
`mind = blown`. Didn't know these text style properties existed -- thanks!

As a side note: I went with `center` instead `middle` for the `dominant-baseline` as the latter was looking uglier due to `p` character that was shifting the middle.

---

<sup>**[src/main/LocationTrack.js, line 75 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-_CVINJh3BjV4kkmD)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L75)):</sup>
`/ 2` was causing it to look like misaligned and that is partly due to our `p` character there. `dominant-baseline` property you suggested fixed also this issue.

---

<sup>**[src/main/Root.js, line 12 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-d3eUKkWimjqy9N3N)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/Root.js#L12)):</sup>
good catch! Removed this require.

---

<sup>**[src/main/ScaleTrack.js, line 2 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-dqGK-QOmu__rV2Ov)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/ScaleTrack.js#L2)):</sup>
I don't know how I managed to get the rest of this deleted, but I blame two monitors and window-focus. Thanks for noticing this one.

---

<sup>**[src/main/ScaleTrack.js, line 56 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-gMdpiocWSReB0hqG)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/ScaleTrack.js#L56)):</sup>
cool! Changed the implementation to leverage these functionalities.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/241)
<!-- Sent from Reviewable.io -->

armish commented:
This is how it looks like after the last round of edits:
![screen recording 2015-07-29 at 06 34 pm](https://cloud.githubusercontent.com/assets/7809/8971668/8d364de8-3620-11e5-8ec0-2c3bd425867d.gif)

danvk commented:
Looks great, Arman! I have a few comments, all small. Free free to merge when they're addressed.

---

Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/LocationTrack.js, line 16 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUV-dixqdO__3dWouG)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/LocationTrack.js#L16)):</sup>
format these like `CONSTANTS_WITH_UNDERSCORES`

---

<sup>**[src/main/LocationTrack.js, line 51 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-a56jKdts9L-DiTcO)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L51)):</sup>
Looks like `getDOMNode` is deprecated and doesn't get added for ES6 classes. You need to use [`React.findDOMNode(component)`][1] instead.

[1]: https://facebook.github.io/react/blog/2015/03/10/react-v0.13.html#new-features

---

<sup>**[src/main/LocationTrack.js, line 63 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-Zl9SqFXVqzwCqfXz)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/LocationTrack.js#L63)):</sup>
Done: https://github.com/hammerlab/pileup.js/issues/243

---

<sup>**[src/main/pileup.js, line 60 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUY2LEzBhH5dhWzDBy)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/pileup.js#L60)):</sup>
What's the error if you forget to specify `data` on a track that requires it?

If it's a confusing error, it might be worth checking for this mistake explicitly.

---

<sup>**[src/main/pileup.js, line 63 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUXGQO_SNkDRjDDPuf)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/pileup.js#L63)):</sup>
drop this

---

<sup>**[src/main/ScaleTrack.js, line 5 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUXO-Hm_V5xtNcsEo3)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/ScaleTrack.js#L5)):</sup>
these sound like great behaviors to test!

---

<sup>**[src/main/ScaleTrack.js, line 62 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUXX7k9M8IGVtXkZkK)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/ScaleTrack.js#L62)):</sup>
You could even pull this out into a static method for easier testing (i.e. without creating a component).

---

<sup>**[src/main/ScaleTrack.js, line 64 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUXg0x6SOnyfNN78Hd)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/ScaleTrack.js#L64)):</sup>
Slightly shorter as `Math.max(1, tmpViewSize)`

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/241)
<!-- Sent from Reviewable.io -->

danvk commented:
BTW, I noticed `Coverage decreased (-2.5%) to 92.174%` from Travis. You could mostly fix this by adding both of these tracks to `src/test/components-test.js`.

armish commented:
ah, didn't know about the `components-tests.js`. Just added them there.

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/pileup.js, line 60 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUY2LEzBhH5dhWzDBy)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/pileup.js#L60)):</sup>
That's a good idea. Here is what we have in the console if user forgets to pass it for a track that doesn't have a default option:

[console error for missing data source](http://i.imgur.com/mJrD3RO.png)

---

<sup>**[src/main/ScaleTrack.js, line 2 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jv-dqGK-QOmu__rV2Ov)** ([raw file](https://github.com/hammerlab/pileup.js/blob/75f265764a332d65fc0b34dcdc1fc7901bb1a58a/src/main/ScaleTrack.js#L2)):</sup>
Done.

---

<sup>**[src/main/ScaleTrack.js, line 5 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUXO-Hm_V5xtNcsEo3)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/ScaleTrack.js#L5)):</sup>
these have slightly changed with the recent changes, but added the relevant cases to the util test.

---

<sup>**[src/main/ScaleTrack.js, line 62 \[r3\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-JvUXX7k9M8IGVtXkZkK)** ([raw file](https://github.com/hammerlab/pileup.js/blob/8c3f641a9656773d1d32106ea2a0dc2d376c48ca/src/main/ScaleTrack.js#L62)):</sup>
this is now part of the utils collection.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/241)
<!-- Sent from Reviewable.io -->

danvk commented:
Looks great -- feel free to merge on green!

---

Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[src/test/utils-test.js, line 143 \[r5\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jvo_P6v0HHKJ_Imcu2U)** ([raw file](https://github.com/hammerlab/pileup.js/blob/20e7c42827cd3455fd5e4f98b4f10f1836c33446/src/test/utils-test.js#L143)):</sup>
trailing whitespace

---

<sup>**[style/pileup.css, line 18 \[r5\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/241#-Jvo_NcVsqslclLD-_kr)** ([raw file](https://github.com/hammerlab/pileup.js/blob/20e7c42827cd3455fd5e4f98b4f10f1836c33446/style/pileup.css#L18)):</sup>
two space indent

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/241)
<!-- Sent from Reviewable.io -->
